src/mirage_net.ml

```diff
@@ -54,8 +54,8 @@ module type S = sig
   val pp_error: error Fmt.t
   type t
   val disconnect : t -> unit Lwt.t
-  val write: t -> size:int -> (Cstruct.t -> int) -> (unit, error) result Lwt.t
-  val listen: t -> header_size:int -> (Cstruct.t -> unit Lwt.t) -> (unit, error) result Lwt.t
+  val write: t -> size:int -> (bytes -> int) -> (unit, error) result Lwt.t
+  val listen: t -> header_size:int -> (string -> unit Lwt.t) -> (unit, error) result Lwt.t
```

Comment:
Or maybe we can try to break even more the API:
```Ocaml
  val write: t -> off:int -> len:int -> string -> (unit, error) result Lwt.t
  val listen: t -> off:int -> len:int -> bytes -> (unit, error) result Lwt.t
```
I'm not sure about the names, for example with the `write` implementations, the callback result is compared to the `size` argument, and now we have to trust it. Similarly for `listen` where `header_size` is now an offset to write the result to, and we have to trust it. As @hannesm pointed out, this boundary checking problem may be now more something about trusting the lower part will work correctly and users will never change that?

Reply:
And with that change, the allocation need to be done in the upper part. :ghost:

Reply:
The issue/reason for write having a callback where you can modify the data is that the network interface should be able to decide which buffer you can write to. As example, the mirage-net-xen implementation uses page-aligned memory, and for it to not need to copy/blit any data from the above layers, it passes the buffer to the client.

Now, that may be wishful thinking, since `bytes` isn't the kind of memory mirage-net-xen can use (it needs non-moving memory since that memory region is passed to the other xen domain, and only when the other says "ok, can be freed" the memory can be reclaimed). [that's at least how I understand it] -- this also means that we need to copy in any case.